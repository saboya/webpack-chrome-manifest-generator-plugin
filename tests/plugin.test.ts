import { vol } from 'memfs'
import { type PermissionTest, WebpackTestHelper } from './util'

process.env.MEMFS_DONT_WARN = '1'

jest.mock('fs')

const permissionTests: PermissionTest[] = [
  {
    testTitle: 'chrome.storage.* requests for "storage" permission',
    permission: 'storage',
    code: 'chrome.storage.sync.get([\'key\'], (result) => { console.log(\'Value currently is \' + result.key); });',
  },
  {
    testTitle: 'navigator.geolocation.* requests for "geolocation" permission',
    permission: 'geolocation',
    code: 'navigator.geolocation.getCurrentPosition(geoSuccess, geoError);',
  },
  {
    testTitle: 'document.execCommand("paste") requests for "clipboardRead" permission',
    permission: 'clipboardRead',
    code: 'document.execCommand("paste");',
  },
  {
    testTitle: 'document.execCommand("cut") requests for "clipboardWrite" permission',
    permission: 'clipboardWrite',
    code: 'document.execCommand("cut");',
  },
  {
    testTitle: 'document.execCommand("copy") requests for "clipboardWrite" permission',
    permission: 'clipboardWrite',
    code: 'document.execCommand("copy");',
  },
]

const permissionTestHelper = async ({ testTitle, permission, code }: PermissionTest) => {
  test(`${testTitle}`, async () => {
    vol.fromJSON({
      '/src/index.js': `
        /*__TYPE__: "content"*/
        /*__MATCHES__: ["https://www.example.com/*"]*/
        /*__RUN_AT__: "document_idle"*/

        ${code}`,
    })

    const { manifestJSON } = await WebpackTestHelper()

    expect(manifestJSON.permissions).toContain(permission)
  })
}

beforeEach(() => {
  vol.reset()
})

describe('ChromeManifestGeneratorPlugin tests', () => {
  test('Manifest is generated correctly for a basic content script', async () => {
    const scriptUrlMatch = 'https://www.example.com/*'

    vol.fromJSON({
      '/src/index.js': `
        /*__TYPE__: "content"*/
        /*__MATCHES__: ["${scriptUrlMatch}"]*/
        /*__RUN_AT__: "document_idle"*/

        console.log('test');`,
    })

    const { manifestJSON } = await WebpackTestHelper()

    expect(manifestJSON.permissions).toContain(scriptUrlMatch)
  })

  test('Manifest is generated correctly for a basic background script', async () => {
    const scriptUrlMatch = 'https://www.example.com/*'

    vol.fromJSON({
      '/src/index.js': `
        /*__TYPE__: "background"*/
        /*__MATCHES__: ["${scriptUrlMatch}"]*/

        console.log('test');`,
    })

    const { manifestJSON } = await WebpackTestHelper()

    expect(manifestJSON.permissions).toContain('background')
  })

  test('Manifest is generated with supplied content_security_policy', async () => {
    const scriptUrlMatch = 'https://www.example.com/*'

    vol.fromJSON({
      '/src/index.js': `
        /*__TYPE__: "background"*/
        /*__MATCHES__: ["${scriptUrlMatch}"]*/

        console.log('test');`,
    })

    const csp = 'script-src \'self\' \'unsafe-eval\'; object-src \'self\''

    const { manifestJSON } = await WebpackTestHelper({ content_security_policy: csp })

    expect(manifestJSON.content_security_policy).toEqual(csp)
  })

  // eslint-disable-next-line @typescript-eslint/no-floating-promises
  permissionTests.map(permissionTestHelper)
})
