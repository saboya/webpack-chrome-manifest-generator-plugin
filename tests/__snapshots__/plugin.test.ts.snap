// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChromeManifestGeneratorPlugin tests Manifest is generated correctly for a basic background script 1`] = `
{
  "/dist/manifest.json": "{
  "manifest_version": 2,
  "name": "Test Extension",
  "short_name": "TestExtension",
  "description": "This is a text extension",
  "version": "1.0.0",
  "permissions": [
    "background"
  ],
  "web_accessible_resources": [
    "img/*"
  ],
  "background": {
    "scripts": [
      "default.js"
    ]
  },
  "content_scripts": []
}",
}
`;

exports[`ChromeManifestGeneratorPlugin tests Manifest is generated correctly for a basic content script 1`] = `
{
  "/dist/manifest.json": "{
  "manifest_version": 2,
  "name": "Test Extension",
  "short_name": "TestExtension",
  "description": "This is a text extension",
  "version": "1.0.0",
  "permissions": [
    "https://www.example.com/*"
  ],
  "web_accessible_resources": [
    "img/*"
  ],
  "background": {
    "scripts": []
  },
  "content_scripts": [
    {
      "js": [
        "default.js"
      ],
      "type": "content",
      "run_at": "document_idle",
      "matches": [
        "https://www.example.com/*"
      ]
    }
  ]
}",
}
`;

exports[`ChromeManifestGeneratorPlugin tests Manifest is generated with supplied content_security_policy 1`] = `
{
  "/dist/manifest.json": "{
  "manifest_version": 2,
  "name": "Test Extension",
  "short_name": "TestExtension",
  "description": "This is a text extension",
  "version": "1.0.0",
  "permissions": [
    "background"
  ],
  "web_accessible_resources": [
    "img/*"
  ],
  "background": {
    "scripts": [
      "default.js"
    ]
  },
  "content_scripts": [],
  "content_security_policy": "script-src 'self' 'unsafe-eval'; object-src 'self'"
}",
}
`;

exports[`ChromeManifestGeneratorPlugin tests chrome.storage.* requests for "storage" permission 1`] = `
{
  "/dist/manifest.json": "{
  "manifest_version": 2,
  "name": "Test Extension",
  "short_name": "TestExtension",
  "description": "This is a text extension",
  "version": "1.0.0",
  "permissions": [
    "storage",
    "https://www.example.com/*"
  ],
  "web_accessible_resources": [
    "img/*"
  ],
  "background": {
    "scripts": []
  },
  "content_scripts": [
    {
      "js": [
        "default.js"
      ],
      "type": "content",
      "run_at": "document_idle",
      "matches": [
        "https://www.example.com/*"
      ]
    }
  ]
}",
}
`;

exports[`ChromeManifestGeneratorPlugin tests document.execCommand("copy") requests for "clipboardWrite" permission 1`] = `
{
  "/dist/manifest.json": "{
  "manifest_version": 2,
  "name": "Test Extension",
  "short_name": "TestExtension",
  "description": "This is a text extension",
  "version": "1.0.0",
  "permissions": [
    "clipboardWrite",
    "https://www.example.com/*"
  ],
  "web_accessible_resources": [
    "img/*"
  ],
  "background": {
    "scripts": []
  },
  "content_scripts": [
    {
      "js": [
        "default.js"
      ],
      "type": "content",
      "run_at": "document_idle",
      "matches": [
        "https://www.example.com/*"
      ]
    }
  ]
}",
}
`;

exports[`ChromeManifestGeneratorPlugin tests document.execCommand("cut") requests for "clipboardWrite" permission 1`] = `
{
  "/dist/manifest.json": "{
  "manifest_version": 2,
  "name": "Test Extension",
  "short_name": "TestExtension",
  "description": "This is a text extension",
  "version": "1.0.0",
  "permissions": [
    "clipboardWrite",
    "https://www.example.com/*"
  ],
  "web_accessible_resources": [
    "img/*"
  ],
  "background": {
    "scripts": []
  },
  "content_scripts": [
    {
      "js": [
        "default.js"
      ],
      "type": "content",
      "run_at": "document_idle",
      "matches": [
        "https://www.example.com/*"
      ]
    }
  ]
}",
}
`;

exports[`ChromeManifestGeneratorPlugin tests document.execCommand("paste") requests for "clipboardRead" permission 1`] = `
{
  "/dist/manifest.json": "{
  "manifest_version": 2,
  "name": "Test Extension",
  "short_name": "TestExtension",
  "description": "This is a text extension",
  "version": "1.0.0",
  "permissions": [
    "clipboardRead",
    "https://www.example.com/*"
  ],
  "web_accessible_resources": [
    "img/*"
  ],
  "background": {
    "scripts": []
  },
  "content_scripts": [
    {
      "js": [
        "default.js"
      ],
      "type": "content",
      "run_at": "document_idle",
      "matches": [
        "https://www.example.com/*"
      ]
    }
  ]
}",
}
`;

exports[`ChromeManifestGeneratorPlugin tests navigator.geolocation.* requests for "geolocation" permission 1`] = `
{
  "/dist/manifest.json": "{
  "manifest_version": 2,
  "name": "Test Extension",
  "short_name": "TestExtension",
  "description": "This is a text extension",
  "version": "1.0.0",
  "permissions": [
    "geolocation",
    "https://www.example.com/*"
  ],
  "web_accessible_resources": [
    "img/*"
  ],
  "background": {
    "scripts": []
  },
  "content_scripts": [
    {
      "js": [
        "default.js"
      ],
      "type": "content",
      "run_at": "document_idle",
      "matches": [
        "https://www.example.com/*"
      ]
    }
  ]
}",
}
`;
